java/arcs/core/data/expression/TypeEvaluator.kt

```diff
@@ -0,0 +1,155 @@
+ * usually a [SeqType] or [ScopeType].
+ *
+ * For [BinaryOp] numeric operations, the LHS and RHS are [widen]'ed to the bigger of the two
+ * types.
+ *
+ * Unlike [ExpressionEvaluator], [Scope] objects here do not contain values, but rather types.
+ */
+class TypeEvaluator(
+  val parameterScope: Scope = ParameterScope(),
+  val scopeCreator: (String) -> Scope = { name -> MapScope<InferredType>(name, mutableMapOf()) }
+) : Expression.Visitor<InferredType, Scope> {
+
+  override fun <E, T> visit(expr: Expression.UnaryExpression<E, T>, ctx: Scope): InferredType {
+    return when (expr.op) {
+      is Expression.UnaryOp.Not -> InferredType.Primitive.BooleanType
+      is Expression.UnaryOp.Negate -> expr.expr.accept(this, ctx)
+    }
+  }
+
+  override fun <L, R, T> visit(
+    expr: Expression.BinaryExpression<L, R, T>,
+    ctx: Scope
+  ): InferredType {
+    return when (expr.op) {
+      // Boolean ops
+      Expression.BinaryOp.And, Expression.BinaryOp.Or,
+      Expression.BinaryOp.LessThan, Expression.BinaryOp.GreaterThan,
+      Expression.BinaryOp.LessThanOrEquals,
+      Expression.BinaryOp.GreaterThanOrEquals,
+      Expression.BinaryOp.Equals,
+      Expression.BinaryOp.NotEquals -> InferredType.Primitive.BooleanType
+
+      // Numeric ops
+      Expression.BinaryOp.Add,
+      Expression.BinaryOp.Subtract,
+      Expression.BinaryOp.Multiply,
+      Expression.BinaryOp.Divide -> widen(
+        expr.left.accept(this, ctx),
+        expr.right.accept(this, ctx)
+      )
+
+      // Special ops
+      Expression.BinaryOp.IfNull -> expr.left.accept(this, ctx).union(expr.right.accept(this, ctx))
+    }
+  }
+
+  override fun <T> visit(expr: Expression.FieldExpression<T>, ctx: Scope): InferredType =
+    (if (expr.qualifier == null) {
+      ctx
+    } else {
+      (expr.qualifier.accept(this, ctx) as? ScopeType)?.scope
+    }).apply {
+      @Suppress("SENSELESS_COMPARISON") if (this == null && !expr.nullSafe) {
+        throw IllegalArgumentException("Field '${expr.field}' not looked up on null scope")
+      }
+    }?.lookup(expr.field) as InferredType
+
+  override fun <E> visit(expr: Expression.QueryParameterExpression<E>, ctx: Scope): InferredType {
+    return parameterScope.lookup(expr.paramIdentifier) as? InferredType
+      ?: throw IllegalArgumentException(
+        "Unbound parameter '${expr.paramIdentifier}'"
+      )
+  }
+
+  override fun visit(expr: Expression.NumberLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.TextLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.BooleanLiteralExpression, ctx: Scope) =
+    expr.value::class.toType()
+
+  override fun visit(expr: Expression.NullLiteralExpression, ctx: Scope) =
+    InferredType.Primitive.NullType
+
+  private fun seqToScopeOr(inferredType: InferredType?, ctx: Scope): Scope =
+    ((inferredType as? SeqType)?.type as? ScopeType)?.scope ?: ctx
```

Comment:
Is there a reason to be so lenient with `as?` instead of `as`? I am worried that we could swallow an error (having a qualifier that is not of type Seq, or not `Seq<Scope>` would be an error I think) and don't notice it.